src/cockpit/remote.py

```diff
@@ -15,21 +15,185 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
+import asyncio
+import getpass
+import logging
+import os.path
+import re
+import socket
 
-from typing import Dict
+from typing import Dict, Optional, Tuple
 
-from .router import RoutingError, RoutingRule
+from ._vendor import ferny
+from .router import RoutingRule, Router
+from .peer import Peer, PeerError
+
+logger = logging.getLogger(__name__)
+
+
+class PasswordResponder(ferny.InteractionResponder):
+    PASSPHRASE_RE = re.compile(r"Enter passphrase for key '(.*)': ")
+
+    password: Optional[str]
+    error_message: Optional[str]
+    attempts: int
+
+    def __init__(self, password: Optional[str]):
+        self.password = password
+        self.error_message = None
+        self.attempts = 0
+
+    async def do_hostkey(self, reason: str, host: str, algorithm: str, key: str, fingerprint: str) -> bool:
+        try:
+            logger.debug('checking ~/.ssh/known_hosts')
+            with open(os.path.expanduser('~/.ssh/known_hosts')) as file:
+                logger.debug('  file is open')
+                if key in file.read():
+                    logger.debug('  and contains the key!')
+                    return False  # this looks like it's going to succeed on its own
+                logger.debug('  but does not contain the key :(')
+        except OSError as exc:
+            logger.debug('  err: %s', exc)
+
+        raise PeerError('unknown-hostkey', error='unknown-hostkey', auth_method_results={},
+                        host_key=f'{host} {algorithm} {key}', host_fingerprint=fingerprint)
+
+    async def do_askpass(self, messages: str, prompt: str, hint: str) -> Optional[str]:
+        logger.debug('Got askpass(%s): %s', hint, prompt)
+
+        match = PasswordResponder.PASSPHRASE_RE.fullmatch(prompt)
+        if match is not None:
+            path = match.group(1)
+            logger.debug("This is a passphrase request for %s, but we don't do those.  Abort.", path)
+            self.error_message = f'locked identity: {path}'
+            return None
+
+        assert self.password is not None
+        assert self.attempts == 0
+        self.attempts += 1
+        return self.password
+
+
+class SshPeer(Peer):
+    session: Optional[ferny.Session] = None
+    host: str
+    user: Optional[str]
+    password: Optional[str]
+    private: bool
+
+    async def do_connect_transport(self) -> asyncio.Transport:
+        assert self.session is not None
+        logger.debug('Starting ssh session user=%s, host=%s, private=%s', self.user, self.host, self.private)
+
+        basename, colon, portstr = self.host.rpartition(':')
+        if colon and portstr.isdigit():
+            host = basename
+            port = int(portstr)
+        else:
+            host = self.host
+            port = None
```

Comment:
This silently ignores cases like `self.host == 'foo:bar'`. Then `colon == ':` but the isdigit() test fails. This case should throw an error, instead of silently ignoring the `:bar` part.

Reply:
I don't agree that the port will be dropped in this case.  It will hit the `else:` branch and set `host = self.host` (ie: the full original name) and `port = None`.

Reply:
Oh, ok - that feels rather strange and unpredictable, but hmm..